File: iq/16/palindrome/is.go
// Copyright 2026 Samvel Khalatyan. All rights reserved.
//
// Use of this source code is governed by a BSD-style
// license that can be found in the LICENSE file.

package palindrome

import (
	"fmt"
	"iter"
)

type Node struct {
	val  int
	next *Node
}

func (node *Node) String() string {
	if node == nil {
		return "<nil>"
	}
	var nn []int
	for node != nil {
		nn = append(nn, node.val)
		node = node.next
	}
	return fmt.Sprint(nn)
}

func NewList(nn ...int) *Node {
	var (
		head *Node
		last *Node
	)
	for _, n := range nn {
		node := &Node{
			val: n,
		}
		switch {
		case head == nil:
			head = node
		default:
			last.next = node
		}
		last = node
	}
	return head
}

func Is(n *Node) bool {
	for leftNode, rightNode := range split(n) {
		if leftNode.val != rightNode.val {
			return false
		}
	}
	return true
}

func split(n *Node) iter.Seq2[*Node, *Node] {
	return func(yield func(*Node, *Node) bool) {
		if n == nil {
			return
		}
		if n.next == nil {
			return
		}
		backward, forward := splitInHalf(n)
		for backward != nil && forward != nil {
			if !yield(backward.node, forward) {
				break
			}
			backward = backward.next
			forward = forward.next
		}
	}
}

type backwardNode struct {
	node *Node
	next *backwardNode
}

func splitInHalf(n *Node) (*backwardNode, *Node) {
	var (
		backward *backwardNode
		forward  = n
	)
	for step := 1; n != nil; n, step = n.next, step+1 {
		switch step % 2 {
		case 0:
			forward = forward.next
		case 1:
			backward = &backwardNode{
				node: forward,
				next: backward,
			}
		}
	}
	return backward, forward
}

File: iq/16/palindrome/is.cc
// Copyright 2026 Samvel Khalatyan. All rights reserved.
//
// Use of this source code is governed by a BSD-style
// license that can be found in the LICENSE file.

//
// NOLINTNEXTLINE
//go:build ignore

#include "iq/16/palindrome/is.h"

#include <cstddef>
#include <memory>
#include <utility>

namespace iq::palindrome {
namespace {

std::pair<std::shared_ptr<Node>, std::shared_ptr<Node>> findMidpoint(
    std::shared_ptr<Node> list) {
  std::shared_ptr<Node> prev;
  std::shared_ptr<Node> next;
  for (std::size_t size = 1; list; list = list->next, size += 1) {
    if (next == nullptr) {
      next = list;
      prev = std::make_shared<Node>(list->val, prev);
      continue;
    }
    if (size % 2 == 0) {
      next = next->next;
    } else {
      prev = std::make_shared<Node>(next->val, prev);
    }
  }
  return std::make_pair(prev, next);
}

}  // namespace

bool Is(const std::shared_ptr<Node>& list) {
  if (list == nullptr) {
    return true;
  }
  if (list->next == nullptr) {
    return true;
  }
  auto [prev, next] = findMidpoint(list);
  while (prev != nullptr && next != nullptr) {
    if (prev->val != next->val) {
      return false;
    }
    prev = prev->next;
    next = next->next;
  }
  return prev == nullptr && next == nullptr;
}

}  // namespace iq::palindrome
